### PR TITLE
Updated sign in callout with flaky test fix

### PR DESCRIPTION
**Pull request by hannahbergam**
This PR updates the callout to allow for sign-in in one click. This required adding a new styling to the sign in button in the application file, which is only added if the callout is visible, and removed on click. The image is also updated to be square again. See below for differences. 
(New):
FancyButton

(Old):
OriginalButton

Additionally, this PR updates a flaky test on iPhone and iPad. The Saucelabs run seemed to indicate that the tests were all happening within the same mobile session. Upon rerun (which starts the test over by itself), the test passed every time. To address this, I forced the test to clear session storage and cookies, ensuring the popup would appear again and the test could operate as expected. However, Saucelabs still has issues connecting with browsers to clear cookies mid-test when on IE and mobile, so have removed these from the tests where this part is necessary.

This required adding an "And I clear session storage" to the steps.rb file. The build was tested on all browsers. 

- jira tickets:
- https://codedotorg.atlassian.net/browse/LP-1770
- https://codedotorg.atlassian.net/browse/LP-1769

### CODE REVIEW

File: apps/src/code-studio/components/header/SignInCallout.jsx [hunk 1]
```diff
@@ -100,11 +103,10 @@ export default class SignInCallout extends React.Component {
 
   render() {
     return (
-      <div style={styles.container}>
+      <div style={styles.container} onClick={this.props.handleClose}>
         <div
           className="login-callout modal-backdrop"
```
Comment (jamescodeorg) [inline apps/src/code-studio/components/header/SignInCallout.jsx]: Do we still need the login-callout class here?  Or maybe it should be on a different element?  (Also, if it's just for testing, we should probably name it `uitest-something`.)

File: apps/src/code-studio/components/header/SignInCallout.jsx [hunk 2]
```diff
@@ -28,8 +28,10 @@ const styles = {
   modalBackdrop: {
     // Most backdrop attributes come from the 'modal-backdrop' class defined by bootstrap
     // but we need to override the opacity as the default opacity of 0.8 is too dark.
-    // Note that bootstrap defaults the z-index of the backdrop to 1040.
-    opacity: 0.5
+    // Setting the z-index to -1 allows it to sit behind the sign in button that we are
+    // pulling on top. It does not keep the backdrop from being clickable.
+    opacity: 0.5,
+    zIndex: -1
```
Comment (jamescodeorg) [inline apps/src/code-studio/components/header/SignInCallout.jsx:34]: For my education, why does -1 work?  I kind of expected a value of -1 to put the backdrop behind all of the other controls on the page (which have positive z-indexes)?

File: dashboard/app/assets/stylesheets/application.scss [hunk 3]
```diff
@@ -334,6 +334,16 @@ img.video_thumbnail {
   #signin_button .header_button {
     margin-top: 6px;
   }
+
+  #sign_in_or_user {
+    &.z_index_above_modal {
+      z-index: 1050;
+
+      .header_user {
```
Comment (jamescodeorg) [inline dashboard/app/assets/stylesheets/application.scss:343]: Just curious, why do we need to specify the background-color here?

File: apps/src/code-studio/components/header/SignInCalloutWrapper.js [hunk 4]
```diff
@@ -24,6 +24,14 @@ export default class SignInCalloutWrapper extends React.Component {
     };
   }
 
+  componentDidMount() {
+    if (!this.state.hideCallout) {
+      this.signInElement = document.getElementById('sign_in_or_user');
```
Comment (jamescodeorg) [inline apps/src/code-studio/components/header/SignInCalloutWrapper.js:31]: Hardcoding this element id here makes me a little uneasy.  How hard would it be to pass a ref (or at least the element-id of the element to highlight?) as a prop?

Btw, should this file have a .jsx extension instead of .js extension?

Comment (jamescodeorg) [inline apps/src/code-studio/components/header/SignInCalloutWrapper.js:31]: (minor) L29-31 and L43-44 aren't immediately obvious as to their intent.  Maybe consider pulling these out into functions with descriptive names or maybe a comment?  Up to you.

Comment (jamescodeorg) [inline apps/src/code-studio/components/header/SignInCalloutWrapper.js:31]: Mostly curious, why did you add the zorder logic here in the wrapper instead of, say, mount and unmount in SignInCallout?

File: dashboard/test/ui/features/star_labs/signin_callout.feature [hunk 5]
```diff
@@ -6,6 +6,23 @@ Scenario: Should see callout on 20-hour farmer lesson
   And I wait for the page to fully load
   And element ".login-callout" is visible
 
+@no_mobile
+@no_ie
+# Issue with cookie clearing on IE
```
Comment (jamescodeorg) [inline dashboard/test/ui/features/star_labs/signin_callout.feature]: Just curious, what's the issue?

File: dashboard/test/ui/features/star_labs/signin_callout.feature [hunk 6]
```diff
@@ -37,7 +54,9 @@ Scenario: See age callout, not signin callout on hour of code
   And I wait until I don't see selector "#p5_loading"
   And I select age 10 in the age dialog
 
-Scenario: After I dismiss the callout, it should not reappear upon refresh
+# Cookies operate on mobile according to device settings: test only passes on rerun
+@no_mobile
```
Comment (jamescodeorg) [inline dashboard/test/ui/features/star_labs/signin_callout.feature:61]: Unrelated to this change, but this reminds me, did we check that the callout looks and works ok on mobile?

Comment (hannahbergam) [reply dashboard/test/ui/features/star_labs/signin_callout.feature:61]: I've been testing on my iPhone, and it looks good!

File: apps/src/code-studio/components/header/SignInCallout.jsx (hunk 2 above)
Comment (hannahbergam) [reply apps/src/code-studio/components/header/SignInCallout.jsx:34]: Maureen was teaching me about this. From what I understand, this ensures the backdrop will show up just behind its parent, which in this case is pulled all the way to the top. @maureensturgeon please correct me if I stated this wrong!

File: dashboard/app/assets/stylesheets/application.scss (hunk 3 above)
Comment (hannahbergam) [reply dashboard/app/assets/stylesheets/application.scss:343]: There isn't usually a color, so the button is transparent and just relies on the color of the teal header behind it. Giving it a color here makes it appear ahead of the backdrop the way we want.

File: apps/src/code-studio/components/header/SignInCalloutWrapper.js (hunk 4 above)
Comment (hannahbergam) [reply apps/src/code-studio/components/header/SignInCalloutWrapper.js:31]: Great question- this was initially because the logic for onClick was here, and it made sense to explicitly remove the extra formatting in the same place we update the cookies/state/etc. But I could go either way on this!

File: dashboard/test/ui/features/star_labs/signin_callout.feature (hunk 5 above)
Comment (hannahbergam) [reply dashboard/test/ui/features/star_labs/signin_callout.feature]: The manual cookie clearing was working in every browser except IE. I can go back and pull this in again to see what the issue was (and then leave a more descriptive comment here).

File: apps/src/code-studio/components/header/SignInCalloutWrapper.js (hunk 4 above)
Comment (hannahbergam) [reply apps/src/code-studio/components/header/SignInCalloutWrapper.js:31]: I had these pulled out in a function but it seemed more complicated than was necessary. I'm happy to add a couple comments to explain!

File: apps/src/code-studio/components/header/SignInCallout.jsx (hunk 1 above)
Comment (hannahbergam) [reply apps/src/code-studio/components/header/SignInCallout.jsx]: I remember why I added this here- we want to specifically test clicks on the modal backdrop, because using the root class that's defined higher up the chain would result in clicks to the sign in button, not the background, and we want to test that clicking the background dismisses it. With that in mind, does it make sense to keep here? If so, I'll definitely update the name.

Comment (hannahbergam) [reply apps/src/code-studio/components/header/SignInCallout.jsx]: Or, I realize I could nest these 3 siblings under a className, but it doesn't seem like good practice to add another div level just for testing.

Comment (jamescodeorg) [reply apps/src/code-studio/components/header/SignInCallout.jsx]: Ah, I see.  If you specifically want to test clicking the modal backdrop, then I think it'd be ok to just reference the `modal-backdrop` class in the test (and remove `login-callout`).

File: apps/src/code-studio/components/header/SignInCallout.jsx (hunk 2 above)
Comment (jamescodeorg) [reply apps/src/code-studio/components/header/SignInCallout.jsx:34]: Oh whoa, I always thought z-orders were absolute.  I didn't realize that z-orders are only considered in the context of the parent frame (so they're only compared against sibling z-orders).  TIL, thanks!

File: dashboard/app/assets/stylesheets/application.scss (hunk 3 above)
Comment (jamescodeorg) [reply dashboard/app/assets/stylesheets/application.scss:343]: Ah ok.  A bit fragile but I don't see a better way.

File: apps/src/code-studio/components/header/SignInCallout.jsx (hunk 1 above)
Comment (hannahbergam) [reply apps/src/code-studio/components/header/SignInCallout.jsx]: @jamescodeorg Okay- will update this now. I realize though that I should test that clicking the box doesn't redirect to sign in (and instead dismisses, just like the backdrop). If I want to test these both explicitly and independently, then is it worth adding a className to the callout div as well?

File: dashboard/app/assets/stylesheets/application.scss (hunk 3 above)
Comment (hannahbergam) [reply dashboard/app/assets/stylesheets/application.scss:343]: I will add a small comment here explaining this!

File: apps/src/code-studio/components/header/SignInCalloutWrapper.js (hunk 4 above)
Comment (hannahbergam) [reply apps/src/code-studio/components/header/SignInCalloutWrapper.js:31]: I can definitely pass it as a prop instead. And thanks for your comment about the filename! I can change it to jsx.

File: dashboard/test/ui/features/star_labs/signin_callout.feature (hunk 5 above)
Comment (hannahbergam) [reply dashboard/test/ui/features/star_labs/signin_callout.feature]: In Saucelabs I get "Failed: Error communicating with the remote browser. It may have died." The cookies work as expected when tested in IE locally, and a rerun always passes. I'm going to keep @no_ie excluded for now.

File: apps/src/code-studio/components/header/SignInCalloutWrapper.js (hunk 4 above)
Comment (hannahbergam) [reply apps/src/code-studio/components/header/SignInCalloutWrapper.js:31]: @jamescodeorg Quick follow up- what makes you want to pass it in as a prop?

Comment (jamescodeorg) [reply apps/src/code-studio/components/header/SignInCalloutWrapper.js:31]: Hannah and I talked about this offline and decided that it wasn't worth the work and risk to pass this in (and change the way the style is applied).

Comment (hannahbergam) [reply apps/src/code-studio/components/header/SignInCalloutWrapper.js:31]: I've updated the file to be .jsx, and it's commented to people can trace it back to the application.scss file updates.

File: dashboard/test/ui/features/step_definitions/steps.rb [hunk 7]
```diff
@@ -1216,8 +1216,8 @@ def find_test_user_by_name(name)
 
 And /^I dismiss the login reminder$/ do
   steps %Q{
-    And I click selector ".login-callout" if I see it
-    And I wait until I don't see selector ".login-callout"
+    And I click selector ".modal-backdrop" if I see it
```
Comment (maureensturgeon) [inline dashboard/test/ui/features/step_definitions/steps.rb:1219]: Is this `modal-backdrop` class used on other components? Is it possible that that you could accidentally dismiss a modal that isn't the signin modal?

Comment (hannahbergam) [reply dashboard/test/ui/features/step_definitions/steps.rb:1219]: @jamescodeorg and I were talking about this for a while and I think we concluded it's okay to test a click to this directly. To answer your question- it is used in a couple other places in LP. I ended up testing clicks to both the backdrop and the callout box itself (even within the same UI test), so figured this variety would help catch any future issues.

Comment (hannahbergam) [reply dashboard/test/ui/features/step_definitions/steps.rb:1219]: @maureensturgeon I'm remembering that naming of the individual components was getting a little messy, and this was the simplest path forward. In the end, I didn't want to name the entire component grouped together (and add an extra div level just for testing), because I wanted to be able to test clicks to both the backdrop and the callout box itself. So the compromise was to move the additional name to the callout box, and then just make sure I tested both pieces. All that said, I'd be happy to put a label on the backdrop too to be safe!

**Review by maureensturgeon — APPROVED** [review]
